### DISCUSSION TITLE
How to use the headless OS and configure mic/speaker/settings

### DISCUSSION THREAD

**Discussion by vgnmnky** [OpenVoiceOS/ovos-buildroot#280] [Q&A] [answered]
I've flashed the headless version of the OS to a Pi 3.  I then booted it, and the wifi hotspot came up and I connected and then configured it with my wifi, and continued.   I can see via my router that it is connected, but what am I supposed to do now to interact with it and make sure the connected devices are recognised/configured?  As far as I can tell, there's no webgui, and no SSH (not that I know what the user and password would be).  The headless documentation just stops dead there.  Of course, I've tried speaking to it ("Hey Mycroft, what time is it?" etc), but there's not a peep back, so I don't know if any of the mic/speaker devices I've tried are correctly configured.

A frustrating experience.  Has anyone else got headless working on a Pi 3?  For info, I've tried with the old AIY voice kit (free with MagPi years ago) and a Pi-top Pulse, both of which have a mic and speaker, and both of which I've had working before, including with Mycroft years ago.  (I initially tried installing it on top of Pi OS lite, with the script, but had issues installing everything it needed).

**Comment by mjlill** [comment, marked as answer]
The Pi3B+  works well with the Google AIY V1 hat (MagPi issue 55)  
#New install of BookWorm 64bit lite
#Burn new image, connect to monitor & kybd for setup
#follow meno for start of local
#Set your user name & add passwd

user  pi
passwd ****create*one*****

#Now login as the new user:
login pi
passwd ****your new pw******

# sudo raspi-config
#set wifi
#set hostname
#turn off blanking 
#enable ssh, spi, (optional i2c, serial (no), hardware (yes))

#set local, tz, keyboard
#Exit NO reboot yet
#check for IP  then reboot
 ifconfig

# Increase swap size for Pi3B+ & Pi4-2g
 sudo nano /etc/dphys-swapfile
# set value to be
CONF_SWAPSIZE=1024

 sudo reboot
 
 # connect ssh or locally into the new Pi  
# (can connect with the <user>@<hostname>.local )
#------------------------------------------------

*Prepare:
 sudo apt update
 sudo apt upgrade -y
 sudo apt install git -y

"""
***** Optional Google AIY V1 kit (original) *****
#Edit config.txt
 sudo nano /boot/firmware/config.txt
add # (to disable for hdmi audio)
#dtparam=audio=on

#Add [all] at bottom of file
dtoverlay=googlevoicehat-soundcard

sudo reboot  #(to activate hat audio)

#Verify only googlevoicehat card 0
 aplay -l 
 arecord -l

"""

# Do OVOS installer:
# copy  & paste /type following command line as per installer 
 sh -c "curl -s https://raw.githubusercontent.com/OpenVoiceOS/ovos-installer/main/installer.sh -o installer.sh && chmod +x installer.sh && sudo ./installer.sh && rm installer.sh"

Accept default settings, optionally is okay to select extra skills
Give a half hour or so to install, it should tell you.

**Reply by mjlill** [reply]
Sry, MagPi issue 57 :)

**Reply by vgnmnky** [reply]
That's very helpful, thank you.  Yes, it's the one I got with issue 57, still have the magazine + box!  Will give this another go today.

**Comment by mjlill** [comment]
There a few other tips that you may like also. One is that you can make the button led go blinky when mycroft is active.
Two is that the button press does work, yetThree the Pi3 doesn't have quite enough omph to use the guiFour a Pi4-2 works very well with it allFive a Pi5 has a hardware difference and the gpio don't work normal (like Pi33 & Pi4)
Six moving sd card to a Pi5, you have to rerun the installer to activate the gui (don't know why)Seven volume is at abt 40% after install, use alsamixer to set for ~80-90% or 
Eight "hey mycroft", "set volume to 90%"Nine I am still learning :)
Also I am running Ollama on the same Pi3, not the fastest but the model tinyllama is good
Glad to help :)
Mike
    On Saturday, November 16, 2024 at 04:52:44 AM EST, vgnmnky ***@***.***> wrote:  
 
 


That's very helpful, thank you. Yes, it's the one I got with issue 57, still have the magazine + box! Will give this another go today.

**Reply by vgnmnky** [reply]
Thanks.  I've got a used Pi 4 2G on order which should make a big difference.

I'll be looking into the button LED, as that will be handy. 

Anyway, I followed your instructions and it's working.  Not well, on this old Pi. but it functions!  So thanks again.

I am having an issue with the volume.  I can set it in alsamixer, but as soon as it speaks, it's loud, and it resets the volume in alsamixer back again.   I've asked Mycroft to set the volume to a lower value, and it says it has, but it always speaks at the same (loud!) volume.  Not sure yet what the solution is to that.   Without a gui I'm a bit in the dark about what my options are, or even what skills are installed (I selected extra).  There are unfortunately a few dead links around that might have been useful.

**Reply by mjlill** [reply]
The gui is not a desktop so is not really helpful other than information.  It is really intended as a touch screen so can pull up menus and other options.  You can sometimes by using a mouse get a pull down but  I have not figured out how to select a menu choice.

When you receive your Pi4-2g, you can move the card over from the Pi3.  The IP will change but you can ssh in using the hostname.local  Then simply rerun the install command line, select all defaults until the last where you select all 3 options, default choice, extra and gui.  This will update existing files and add the gui. It goes quickly as most files are already installed and is just going to add the gui files.  

Just be patient as it will take several minutes to show the display gui as it loads the skills. Once is dis[played, the 'hey mycroft" will draw a band around the other display edge to show it heard you so there is visual feedback.
You can see the list of skills  with  ls /tmp  but that is just a nose count with nothing of use.

As for the audio, not sure how it does the settings. If it hears the wakeword it will duct the level so will hear the command / request. After several adjustments it seems to self remember what the setting should be.  Use specific values  to adjust the level ie, "hey mycroft", "set volume 65%"   up & down also work but not sure size of adjustments.

To make the led go blinky blinky you will need to install a skill. The push button is still an in work task. (above my level)  That is a fairly easy process to install.

**Reply by mjlill** [reply]
To make the button light up
# https://github.com/smartgic/ovos-phal-plugin-ww-led
git clone https://github.com/smartgic/ovos-phal-plugin-ww-led.git
cd ovos-phal-plugin-ww-led
~/.venvs/ovos/bin/pip install ovos-phal-plugin-ww-led
# no changes, future wip
nano .config/OpenVoiceOS/ovos-phal-plugin-ww-led.json  #(copy & paste)
{
  "gpio_pin": 25,
  "wakeword_only": false,
  "pulse": true
}

**Reply by JarbasAl** [reply]
> As for the audio, not sure how it does the settings. If it hears the wakeword it will duct the level so will hear the command / request. 

this is a config option you can disable if you dont like it https://github.com/OpenVoiceOS/ovos-config/blob/dev/ovos_config/mycroft.conf#L439

**Reply by mjlill** [reply]
For me the ducting is good, would like to make the AIY push button trigger the waleword detected so one could stop a media event :)  (working on that one)

**Reply by vgnmnky** [reply]
I just can't get it to change the volume.  It says it has.  I ask it to set it to whatever, then ask it what it's set to.  Then decrease it, and ask again, and it says the right things.  But the actual volume is always the same level of very loud.  1% sounds the same as 100%.

Got the button LED working, which is nice :)  Thank you.

I've listed the skills in /tmp, but I'm not sure that all of them are actually installed/working.  Tested a few, and it doesn't know what I'm talking about (I looked up the Git pages for them to make sure I was saying the right things).  For example, the parrot one, and a couple of the radio ones.

I suppose I'm not bothered about a gui really, I prefer headless, but a webgui would be cool.

Still, I'm having some fun with it.

**Reply by mjlill** [reply]
"hey mycroft", "lower volume"     should reduce by 10% and anounce the setting

**Reply by vgnmnky** [reply]
It's very odd, I used those commands.  I say "set the volume to 5%" and it says, loudly, "VOLUME ABATED TO 5%".  Then I say "what is the volume", and it says (with the same loudness) "THE VOLUME IS AT 40".  Same when I say set to 100 (max), and ask what the volume is, it still says 40.  Something is resetting/overriding the setting.

**Reply by mjlill** [reply]
This is audio from the AIY?  If setting alsamixer to a lower level it should speak quieter.
With alsamixer open you will see any level changes 
"hey mycroft", " make me laugh"

**Comment by JarbasAl** [comment]
do you have multiple speakers/soundcards? it kinda seems OVOS is changing the wrong sink

what does `aplay -l` show?

since you are using a hat, maybe disable snd_bcm2835 (the rpi audio jack) completley to ensure it isnt selected?

**Reply by vgnmnky** [reply]
It sounded like that to me as well, until I realised that the volume was changing back to 40 whenever it was about to speak (I watched alsamixer while this happened).  aplay -l only shows the HAT, nothing else.

It's possible I messed something up in .asoundrc or something like that, though I did most of the messing yesterday before I started again with a fresh SD today and followed @mjlill  's instructions.

I have to admit, I get a bit confused with the whole Linux audio ecosystem, pulseaudio, pipewire, alsa, etc, though as far as I know I haven't been messing with anything outside alsa.

**Reply by mjlill** [reply]
Just stay with the steps out lined. No changes to the basic OS except as the setup goes through. I have done this several times, Pi3, Pi4 & Pi5 with no audio issues for control. It also hears me from across the room in a normal voice 
For the AIY V1
aplay -l    #card 0: sndrpigooglevoi 
arecord -l #card 0: sndrpigooglevoi 
ls   /etc/asound.conf  #No such file or directory

card 0 is the default for alsa
 less .asoundrc   #as set at install by ovos
 pcm.!default pipewire
 ctl.!default pipewire

When you say "hey mycroft"  the output level is suppose to drop and then return to previous volume level setting when it speaks the response.

**Reply by mjlill** [reply]
Also
 sudo nano /boot/firmware/config.txt
add # (to disable for hdmi audio)
#dtparam=audio=on

Adding the hash to the dtparm disables other audio outputs
